### PR TITLE
feat: Add referral code, events, NatSpec, README (SC-448)

### PR DESCRIPTION
**Pull request by lucas-manuel**
Adds 
- Referral codes to deposit, withdraw, swap
- Adds full natspec documentation
- Adds events and events tests
- Adds README

### CODE REVIEW

**Review by barrutko — DISMISSED** [review]
LGTM

File: src/PSM.sol [hunk 1]
```diff
@@ -65,7 +65,8 @@ contract PSM {
         address assetOut,
         uint256 amountIn,
         uint256 minAmountOut,
-        address receiver
+        address receiver,
+        uint16  referralCode
```
Comment (hexonaut) [inline src/PSM.sol:69]: Let's extend this to `uint256`. I'm not sure why Aave made it smaller (does it save gas..?), but we don't have to blindly duplicate them here.

File: src/PSM.sol [hunk 2]
```diff
@@ -5,20 +5,24 @@ import { IERC20 } from "erc20-helpers/interfaces/IERC20.sol";
...
 }
 
-// TODO: Add events and corresponding tests
 // TODO: Determine what admin functionality we want (fees?)
 // TODO: Refactor into inheritance structure
-// TODO: Add interface with natspec and inherit
 // TODO: Prove that we're always rounding against user
 // TODO: Add receiver to deposit/withdraw
-contract PSM {
+contract PSM is IPSM {
```
Comment (hexonaut) [inline src/PSM.sol:18]: Missing `override` keyword on a bunch of functions and variables defined in the interface.

File: src/PSM.sol [hunk 3]
```diff
@@ -78,25 +79,29 @@ contract PSM {
...
+        emit Swap(assetIn, assetOut, msg.sender, receiver, amountIn, amountOut, referralCode);
     }
 
     /**********************************************************************************************/
     /*** Liquidity provision functions                                                          ***/
     /**********************************************************************************************/
 
-    function deposit(address asset, uint256 assetsToDeposit)
-        external returns (uint256 newShares)
+    function deposit(address asset, uint256 assetsToDeposit, uint16 referralCode)
```
Comment (hexonaut) [inline src/PSM.sol:90]: Don't need referral for LPs.

File: src/PSM.sol [hunk 4]
```diff
@@ -78,25 +79,29 @@ contract PSM {
...
         totalShares        += newShares;
 
         IERC20(asset).safeTransferFrom(msg.sender, address(this), assetsToDeposit);
+
+        emit Deposit(asset, msg.sender, assetsToDeposit, newShares, referralCode);
     }
 
-    function withdraw(address asset, uint256 maxAssetsToWithdraw)
-        external returns (uint256 assetsWithdrawn)
+    function withdraw(address asset, uint256 maxAssetsToWithdraw, uint16 referralCode)
```
Comment (hexonaut) [inline src/PSM.sol:103]: Don't need referral for LPs.

File: src/PSM.sol [hunk 5]
```diff
@@ -171,22 +178,24 @@ contract PSM {
     /*** Swap preview functions                                                                 ***/
```
Comment (hexonaut) [inline src/PSM.sol:178]: Just noticed this is duplicate. Change to `Conversion functions`

File: test/Events.t.sol [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    address sender   = makeAddr("sender");
+    address receiver = makeAddr("receiver");
+
+    function test_deposit_events() public {
+        vm.startPrank(sender);
+
+        dai.mint(sender, 100e18);
+        dai.approve(address(psm), 100e18);
+
+        vm.expectEmit();
```
Comment (hexonaut) [inline test/Events.t.sol:45]: The first argument is of the contract emitter. It's best practise to set this for robustness. IE `vm.expectEmit(address(psm));`. Do this for all instances.

**Review by hexonaut — CHANGES_REQUESTED** [review]
A few changes. Also, unrelated to this particular PR, but wanted to say that we should probably rename this contract to `PSM3` to indicate it is a 3-way PSM. We may add a 2-way one later.

File: src/PSM.sol (hunk 1 above)
Comment (lucas-manuel) [reply src/PSM.sol]: Updated

File: src/PSM.sol (hunk 2 above)
Comment (lucas-manuel) [reply src/PSM.sol:18]: Fixed

File: src/PSM.sol (hunk 3 above)
Comment (lucas-manuel) [reply src/PSM.sol]: Removed

File: src/PSM.sol (hunk 4 above)
Comment (lucas-manuel) [reply src/PSM.sol]: REmoved

File: src/PSM.sol (hunk 5 above)
Comment (lucas-manuel) [reply src/PSM.sol]: Updated

File: test/Events.t.sol (hunk 6 above)
Comment (lucas-manuel) [reply test/Events.t.sol:43]: Nice updated

File: src/PSM.sol (hunk 2 above)
Comment (hexonaut) [reply src/PSM.sol:18]: Still missing override on variables. All things mentioned in the interface should have override on them.

**Review by hexonaut — CHANGES_REQUESTED** [review]
Still one unresolved issue with overrides then good to go.

**Pull request merged**